### PR TITLE
MB-7290-sit-extension-SO

### PR DESCRIPTION
**Pull request by NamibiaTorres**
## Description

Add Service Object for CreateSITExtension

## Reviewer Notes

Is there anything you would like reviewers to give additional scrutiny?

## Setup

To test the code:

```go
go test ./pkg/services/sit_extension --testify.m TestValidationRules -v
```
```go
go test ./pkg/services/sit_extension --testify.m TestSITExtensionCreator -v
```
## References

* [Jira story](https://dp3.atlassian.net/browse/MB-7290) for this change

### CODE REVIEW

File: pkg/services/sit_extension/rules.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+	"github.com/gofrs/uuid"
+)
+
+// checkShipmentID checks that the user can't change the shipment ID
+func checkShipmentID() sitExtensionValidator {
+	return sitExtensionValidatorFunc(func(_ appcontext.AppContext, sitExtension models.SITExtension, _ *models.MTOShipment) error {
+		verrs := validate.NewErrors()
+
+		if sitExtension.MTOShipmentID == uuid.Nil {
```
Comment (eamahanna) [inline pkg/services/sit_extension/rules.go:20]: Should we also check that the shipment ID exists in the DB?

File: pkg/services/sit_extension/rules.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+	"github.com/transcom/mymove/pkg/appcontext"
+	"github.com/transcom/mymove/pkg/models"
+	"github.com/transcom/mymove/pkg/services"
+
+	"fmt"
+
+	"github.com/gofrs/uuid"
+)
+
+// checkShipmentID checks that the user can't change the shipment ID
```
Comment (eamahanna) [inline pkg/services/sit_extension/rules.go:15]: This comment doesn't quite make sense. The shipment ID is required to make a sit extension request.

File: pkg/services/sit_extension/rules.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+		verrs := validate.NewErrors()
+		var sitStatus models.SITExtensionStatus
+		var sitExtensionReason models.SITExtensionRequestReason
+		var sitRequestedDays int
+
+		sitStatus = sitExtension.Status
+		sitExtensionReason = sitExtension.RequestReason
+		sitRequestedDays = sitExtension.RequestedDays
+
+		// Check that we have something in the Status field:
```
Comment (eamahanna) [inline pkg/services/sit_extension/rules.go:41]: Update comment

File: pkg/services/sit_extension/rules_test.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+		err := checkPrimeAvailability(checker).Validate(appcontext.NewAppContext(suite.DB(), suite.logger), models.SITExtension{}, nil)
+		suite.NotNil(err)
+		suite.IsType(services.NotFoundError{}, err)
+		suite.Equal(fmt.Sprintln("not found while looking for Prime-available Shipment"), err.Error())
+	})
+
+	suite.Run("checkPrimeAvailability - Success", func() {
+		currentTime := time.Now()
+		shipment := testdatagen.MakeMTOShipment(suite.DB(), testdatagen.Assertions{
+			Move: models.Move{
```
Comment (eamahanna) [inline pkg/services/sit_extension/rules_test.go:72]: You can always use the func `testdatagen.MakeAvailableMove` to make the move and pass it all in or you will need to add `Status: models.MoveStatusAPPROVED`

**Review by sandy-wright — CHANGES_REQUESTED** [review]
Looks really good! Super solid overall, and I really liked the way you reorganized the `validation.go` file. Subtle, but putting the interface at the bottom helps differentiate it, I think.

I did request a few changes though. Some of them might be confusing (sorry, I was struggling to word some things 😓), so just let me know if you want to go over anything!

File: pkg/services/sit_extension/sit_extension_creator.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+	"fmt"
+)
+
+type sitExtensionCreator struct {
+	checks []sitExtensionValidator
+}
+
+// NewSitExtensionCreator creates a new struct with the service dependencies
+func NewSitExtensionCreator(db *pop.Connection) services.SITExtensionCreator {
```
[the remark this answers is not in the document]
Comment (NamibiaTorres) [reply pkg/services/sit_extension/sit_extension_creator.go:19]: > Should remove this `db` from the function arguments, we don't use it anymore.

Oh yes! I had to do this reweighs too.

File: pkg/services/sit_extension/sit_extension_creator.go [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+			checkShipmentID(),
+			checkRequiredFields(),
+		},
+	}
+}
+
+//// CreateSITExtensionCheck passes teh Prime validator to CreateSITExtension
+//func (f *sitExtensionCreator) CreateSITExtensionCheck(appCtx appcontext.AppContext, sitExtension *models.SITExtension) (*models.SITExtension, error) {
+//	return f.CreateSITExtension(appCtx, sitExtension, f.checks...)
+//}
```
Comment (sandy-wright) [inline pkg/services/sit_extension/sit_extension_creator.go:28]: You can delete this commented out bit entirely now :)

**Review by sandy-wright — APPROVED** [review]
Looks like you need to resolve some conflicts before merging, but it looks good to go! I left one more small comment about deleting some commented-out code, but that was it ☺️. Good work!

**Pull request merged**